src/rvsdg/egglog_optimizer/loop_invariant.egg

```diff
@@ -0,0 +1,168 @@
+(ruleset loop_inv_detect)
+
+;; find invariant
+(relation find_inv_expr (Body Expr))
+(relation find_inv_oprd (Body Operand))
+(relation find_inv_body (Body Body))
+(relation find_inv_vo (Body VecOperand))
+
+;; i64: 0 means the existing term is not invariant, 1 mean it is.
+(function is_inv_expr (Body Expr) i64 :unextractable :merge (max old new))
+(function is_inv_oprd (Body Operand) i64 :unextractable :merge (max old new))
+(function is_inv_body (Body Body) i64 :unextractable :merge (max old new))
+(function is_inv_vo (Body VecOperand) i64 :unextractable :merge (max old new))
+ 
+(relation arg_inv (Body i64))
+;; VecOperand: output of Theta node.
+;; If the node link directly to input, the input operand is invariant.
+;; Body : context
+;; i64: iterator index.
+(relation arg_inv_helper (VecOperand Body i64))
```

Comment:
Put a TODO here- I think can replace this with @rtjoa's idea of having a table with a range of integers